lib/phoenix/endpoint/cowboy_websocket.ex

```diff
@@ -1,21 +1,46 @@
-defmodule Phoenix.Endpoint.CowboyWebsocket do
+defmodule Phoenix.Endpoint.CowboyWebSocket do
+  # Implementation of the WebSocket transport for Cowboy.
   @moduledoc false
+
   @behaviour :cowboy_websocket_handler
+  @connection Plug.Adapters.Cowboy.Conn
+  @already_sent {:plug_conn, :sent}
+
+  def init({transport, :http}, req, {module, opts}) when transport in [:tcp, :ssl] do
+    conn = @connection.conn(req, transport)
+    case module.init(conn, opts) do
+      {:ok, %{adapter: {@connection, req}}, args} ->
+        {:upgrade, :protocol, __MODULE__, req, args}
+      {:error, %{adapter: {@connection, req}}} ->
+        {:shutdown, req, :no_state}
+    end
+  after
+    receive do
+      @already_sent -> :ok
+    after
+      0 -> :ok
+    end
+  end
 
-  def call(conn, args) do
-    resume(conn, :cowboy_websocket, :upgrade, args)
+  def upgrade(req, env, __MODULE__, {handler, opts}) do
+    args = [req, env, __MODULE__, {handler, opts}]
+    resume(:cowboy_websocket, :upgrade, args)
   end
 
-  def resume(conn, module, fun, args) do
+  def terminate(_reason,  _req, _state) do
+    :ok
+  end
+
+  def resume(module, fun, args) do
     try do
       apply(module, fun, args)
     catch
       kind, [{:reason, reason}, {:mfa, _mfa}, {:stacktrace, stack} | _rest] ->
         reason = format_reason(kind, reason, stack)
-        exit({reason, {__MODULE__, :call, [conn, []]}})
+        exit({reason, {__MODULE__, :resume, [module, fun, args]}})
```

Comment:
By using `{__MODULE__, :resume, [module, fun, args]}` we keep the `args` around, which includes the initial `req` and other things. Those arguments probably aren't that helpful to log in the error and we don't need to hold on to them. Instead we could recreate the real mfa, using `:mfa`, `:message`, `:req` and `:state` from the keyword list.

(Un/)Fortunately `:req` is a keyword list and not the actual `cowboy_req` record. However we could turn it back into the record or use the keyword to create a `Cowboy.Request` struct that could implement the inspect protocol.

Thoughts?